File: leetcode/graph/bfs/1091_shortest_path_binary_matrix.py
# LC 1091. Shortest Path in Binary Matrix | Medium
from collections import deque

class Solution:
    def shortestPathBinaryMatrix(self, grid):
        n = len(grid)
        if grid[0][0] or grid[n-1][n-1]: return -1
        q = deque([(0, 0, 1)]); grid[0][0] = 1
        while q:
            i, j, d = q.popleft()
            if i == n-1 and j == n-1: return d
            for di in (-1,0,1):
                for dj in (-1,0,1):
                    ni, nj = i+di, j+dj
                    if 0 <= ni < n and 0 <= nj < n and not grid[ni][nj]:
                        grid[ni][nj] = 1; q.append((ni, nj, d+1))
        return -1

File: leetcode/graph/bfs/1091_shortest_path_binary_matrix.cpp
// LC 1091. Shortest Path in Binary Matrix | Medium
#include <vector>
#include <queue>
using namespace std;

class Solution {
public:
    int shortestPathBinaryMatrix(vector<vector<int>>& grid) {
        int n = grid.size();
        if (grid[0][0] || grid[n-1][n-1]) return -1;
        queue<tuple<int,int,int>> q; q.push({0,0,1}); grid[0][0] = 1;
        while (!q.empty()) {
            auto [i,j,d] = q.front(); q.pop();
            if (i == n-1 && j == n-1) return d;
            for (int di = -1; di <= 1; di++)
                for (int dj = -1; dj <= 1; dj++) {
                    int ni = i+di, nj = j+dj;
                    if (ni >= 0 && ni < n && nj >= 0 && nj < n && !grid[ni][nj]) {
                        grid[ni][nj] = 1; q.push({ni,nj,d+1});
                    }
                }
        }
        return -1;
    }
};
